Sources/Algorithms/Pad.swift

```diff
@@ -0,0 +1,156 @@
+  /// to `paddedCount` by repeating the padding `element` at the end.
+  ///
+  /// In case the length of the collection is longer than or equal to the
+  /// `paddedCount`, the original collection is returned.
+  ///
+  /// This example uses `paddingEnd(with:toCount:)` on `Data` to expand
+  /// it to 8 bytes in length.
+  ///
+  ///     let data = Data([0xDE, 0xAD, 0xBE, 0xEF])
+  ///         .paddingEnd(with: UInt8.zero, toCount: 8)
+  ///     // data contains [0xDE, 0xAD, 0xBE, 0xEF, 0x00, 0x00, 0x00, 0x00]
+  ///
+  /// - Parameters:
+  ///   - element: The element which is repeated to expand the collection.
+  ///   - paddedCount: The length of the padded collection.
+  ///
+  /// - Complexity: O(_m_), when the collection's length is less than the
+  /// `paddedCount`, where _m_ is the `paddedCount`. O(_n_) when the
+  /// collection's length is greater than or equal to the `paddedCount`, where
+  /// _n_ is the length of the collection. For a `RandomAccessCollection`
+  /// when the collection's length is greater than or equal to the `paddedCount`,
+  /// the complexity is reduced to O(_1_).
+  @inlinable
+  public func paddingEnd(
+    with element: Element,
+    toCount paddedCount: Int
+  ) -> Self {
+    let padElementCount = paddedCount - count
+
+    // Early exit if no padding is required.
+    guard padElementCount > 0 else { return self }
+
+    var paddedCollection = Self.init()
+    paddedCollection.reserveCapacity(paddedCount)
+    paddedCollection.append(contentsOf: self)
+    paddedCollection.append(contentsOf: repeatElement(element, count: padElementCount))
+
+    return paddedCollection
+  }
+
+  /// Mutates a `RangeReplaceableCollection` by expanding its length
+  /// to `paddedCount` repeating the padding `element` at the end.
+  ///
+  /// The collection is not mutated in the case where the length of the collection
+  /// is greater than or equal to the `paddedCount`.
+  ///
+  /// This example uses `padEnd(with:toCount:)` to get a `String`
+  /// justified to the desired `paddedCount`.
+  ///
+  ///     var myString = "Hello, World!"
+  ///     myString.padEnd(with: " ", toCount: 20)
+  ///     myString // "Hello, World!       "
+  ///
+  /// - Parameters:
+  ///   - element: The element which is repeated to expand the `Collection`.
+  ///   - paddedCount: The length of the padded `Collection`.
+  ///
+  /// - Complexity: O(_m_), when the collection's length is less than the
+  /// `paddedCount`, where _m_ is the `paddedCount`. O(_n_) when the
+  /// collection's length is greater than or equal to the `paddedCount`, where
+  /// _n_ is the length of the collection. For a `RandomAccessCollection`
+  /// when the collection's length is greater than or equal to the `paddedCount`,
+  /// the complexity is reduced to O(_1_).
+  @inlinable
+  public mutating func padEnd(
+    with element: Element,
+    toCount paddedCount: Int
+  ) {
+    self = paddingEnd(with: element, toCount: paddedCount)
```

Comment:
This is more a question but since this is mutating, would be more efficient just to `self.append(contentsOf: repeatElement(element, count: paddedCount - count))`. I'm just asking because it seems like call the no mutating version and assign to self will allocate a whole new buffer and re-append all elements and I'm not sure if this is really necessary. I didn't read all the discussion collapsed, so sorry if this was already pointed out =]

Reply:
One could make this case for `padStart` as well, but correct me if I'm wrong, insert `n` at begging adds more complexity to most of the collections since all the elements had to be slided `n` positions.

Reply:
I decided to do some performance tests. The tests had to capture both the cases where `capacity <= paddedCount` and `capacity > paddedCount`.

To achieve this, I used the following code to perform the benchmark:

```swift
benchmark.add(
  title: "padStart + 10",
  input: [Int].self
) { input in
  return { timer in
    var inp = input
    inp.reserveCapacity(inp.count * 2)
    timer.measure {
      inp.padStart(with: 0, toCount: inp.count + 10)
    }
  }
}

benchmark.add(
  title: "padStart * 3",
  input: [Int].self
) { input in
  return { timer in
    var inp = input
    inp.reserveCapacity(inp.count * 2)
    timer.measure {
      inp.padStart(with: 0, toCount: inp.count * 3)
    }
  }
}
```

In one case I add a constant amount of padding, and in the other case I am padding the array to 3 times its original length.

The int array input provided by the `swift-collections-benchmark` library always has the same capacity as its length. To aid testing, I reserve 2x the initial capacity (simulating geometric growth). Now, for the case of constant padding, as the size of the input array increases, we would not have to allocate a new buffer. For 3x padding case, we will always allocate a new buffer.

Here are the results:

### padEnd

<img width="1280" alt="chart padend2" src="https://user-images.githubusercontent.com/7317008/129034493-beb55f33-4e6f-4077-9089-b4359cda82aa.png">

implementation of `padEnd2`:

```swift
  mutating func padEnd2(
    with element: Element,
    toCount paddedCount: Int
  ) {
    let padElementCount = paddedCount - count

    // Early exit if no padding is required.
    guard padElementCount > 0 else { return }

    self.append(contentsOf: repeatElement(element, count: padElementCount))
  }
```

**Summary:** Results are unsurprising, both algorithms performs the same when a new buffer is allocated, however `padEnd2` has a constant time (as padding length is fixed in this test case; the try complexity should be dependent on the padding length; padding length = paddedCount - collection.count) when no new buffer allocation is required.

This is a good candidate for optimization.

### padStart

<img width="1280" alt="chart padstart2" src="https://user-images.githubusercontent.com/7317008/129035247-d2cd01d5-400f-4bac-ba57-f50c1add99f5.png">

implementation of `padStart2`:

```swift
  public mutating func padStart2(
    with element: Element,
    toCount paddedCount: Int
  ) {
    let padElementCount = paddedCount - count

    // Early exit if no padding is required.
    guard padElementCount > 0 else { return }

    insert(contentsOf: repeatElement(element, count: padElementCount), at: startIndex)
  }
```

**Summary:** Surprisingly `padStart2` performs better in the case where no buffer allocation is required. The overall complexity of both algorithms should be the same. In the case where a new buffer allocation is required, both algorithms perform similarly.

Based on these results, I guess we can change out the implementation for `padStart` as well. Though I am not sure if these results will hold for all types of `RangeReplaceableCollection`s. What do you think?

<hr>

I believe due to how I was performing the performance tests, the results has more deviation than normal. However, after running the tests several times I always see a similar trend.

Reply:
Those benchmarks were nice, to me `padEnd` is definitely worth it, but maybe the guys from the swift team have some other considerations? 

> Based on these results, I am not sure if we have a clear answer here. What do you think?

For `padStart` this is somewhat expected as well, and I agree that is not a clear answer. I think  is good that for small padElementCount there is an improvement ... but if is worth it or not because of the difference on `* 3`. A hybrid version where choosing between one or another based on `padElementCount` would be an option? Not really sure.

Reply:
Very interesting benchmarks, especially the `padStart` one. While the time complexity of reusing the current buffer and allocating a new buffer is the same, I wonder if in the latter case an overwhelming portion of the total runtime goes towards the allocation itself which `insert(contentsOf:at:)` can (sometimes) avoid.

Either way, both of these changes look good to me. The slight slowdown in certain cases doesn't concern me too much, because
* it only appears starting at a few hundred elements which I don't think is a typical use case for these methods, and
* there's no reason I can think of why inserting elements at the start should be slower than putting it into a new buffer manually — this looks more like an optimization `insert(contentsOf:at:)` could make than something `padStart` should need to worry about.

Reply:
>A hybrid version where choosing between one or another based on padElementCount would be an option?

The only way to allow a hybrid approach would be to know the capacity before hand and switch based on if an allocation is required. A `RangeReplaceableCollection` allows you reserve capacity but asking for a collection's capacity is not part of the protocol (presumably not all `RangeReplaceableCollection`s have a notion of capacity, the default implementation of reserveCapacity is empty).

Reply:
>I wonder if in the latter case an overwhelming portion of the total runtime goes towards the allocation itself which insert(contentsOf:at:) can (sometimes) avoid.

Looks like it. In the original implementation, an entire new buffer is allocated. `insert(contentsOf:at:)` should be able to work around it by shifting all the elements in the buffer, as in the case of `padStartX + 10` we know that buffer has enough capacity in all cases for element count > 0 (where `padStart2 + 10` starts diverging).

Reply:
I guess I will change out implementations for both the mutating variants. (The complexity description for padEnd is gonna be interesting to write as there are 4 different trends based on various conditions 🤷)

Reply:
> I guess I will change out implementations for both the mutating variants. (The complexity description for padEnd is gonna be interesting to write as there are 4 different trends based on various conditions 🤷)

Usually costs associated with allocating new memory in case the capacity is insufficient aren't mentioned very explicitly, so don't worry about that too much.

Reply:
For padEnd I have the following 4 cases

General case:
- O(m); count < paddedCount
- O(n); count >= paddedCount

For RandomAccessCollections:
- O(m-n); count < paddedCount
- O(1); count >= paddedCount

m = paddedCount
n = collection.count

Reply:
Ah yeah. You could possibly condense this by defining `m` to be the number of elements that gets added instead: that would make it `O(m)` for random-access collections and `O(n + m)` otherwise, and that also makes it a bit more intuitive what the extra `n` term comes from.